lxd/api_cluster.go

```diff
@@ -2019,60 +2019,126 @@ func clusterCertificatePut(d *Daemon, r *http.Request) response.Response {
 		return response.BadRequest(fmt.Errorf("Private key must be base64 encoded PEM key: %w", err))
 	}
 
+	err = updateClusterCertificate(r.Context(), d, r, req)
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	requestor := request.CreateRequestor(r)
+	s.Events.SendLifecycle(projectParam(r), lifecycle.ClusterCertificateUpdated.Event("certificate", requestor, nil))
+
+	return response.EmptySyncResponse
+}
+
+func updateClusterCertificate(ctx context.Context, d *Daemon, r *http.Request, req api.ClusterCertificatePut) error {
+	revert := revert.New()
+	defer revert.Fail()
+
+	newClusterCertFilename := shared.VarPath("cluster.crt.new")
+
 	// First node forwards request to all other cluster nodes
 	if !isClusterNotification(r) {
+		oldCertBytes, err := os.ReadFile(shared.VarPath("cluster.crt"))
+		if err != nil {
+			return err
+		}
+
+		keyBytes, err := os.ReadFile(shared.VarPath("cluster.key"))
+		if err != nil {
+			return err
+		}
+
+		oldReq := api.ClusterCertificatePut{
+			ClusterCertificate:    string(oldCertBytes),
+			ClusterCertificateKey: string(keyBytes),
+		}
+
 		// Get all members in cluster.
 		var members []db.NodeInfo
-		err = d.db.Cluster.Transaction(r.Context(), func(ctx context.Context, tx *db.ClusterTx) error {
+		err = d.db.Cluster.Transaction(ctx, func(ctx context.Context, tx *db.ClusterTx) error {
 			members, err = tx.GetNodes(ctx)
 			if err != nil {
 				return fmt.Errorf("Failed getting members: %w", err)
 			}
 
 			return nil
 		})
-		if err != nil {
-			return response.SmartError(err)
-		}
 
 		localClusterAddress := d.State().LocalConfig.ClusterAddress()
 
+		revert.Add(func() {
+			// If distributing the new certificate fails, store the certificate. This new file will
+			// be considered when running the auto renewal again.
+			err = os.WriteFile(newClusterCertFilename, []byte(req.ClusterCertificate), 0600)
+			if err != nil {
+				logger.Error("Failed storing new certificate", logger.Ctx{"err": err})
+			}
+		})
+
+		newCertInfo, err := shared.KeyPairFromRaw([]byte(req.ClusterCertificate), []byte(req.ClusterCertificateKey))
+		if err != nil {
+			return err
+		}
+
 		for _, member := range members {
 			if member.Address == localClusterAddress {
 				continue
 			}
 
 			client, err := cluster.Connect(member.Address, d.endpoints.NetworkCert(), d.serverCert(), r, true)
 			if err != nil {
-				return response.SmartError(err)
+				return err
 			}
 
 			err = client.UpdateClusterCertificate(req, "")
 			if err != nil {
-				return response.SmartError(err)
+				return err
 			}
+
+			// When reverting the certificate, we need to connect to the cluster members using the
+			// new certificate otherwise we'll get a bad certificate error.
+			revert.Add(func(address string) func() {
+				return func() {
```

Comment:
This nested function creation seems a little over complicated, when an easier approach to prevent the revert using the wrong variable in the for loop is to use this pattern:

```
for i := range members {
    member := i
    ...
}
```

Reply:
Hopefully we won't have to worry about this too much longer:

https://github.com/golang/go/discussions/56010